server/src/main/java/com/google/sps/utils/StateCollectionBuilder.java

```diff
@@ -0,0 +1,103 @@
+ *              document contains a ads field, which is an ArrayList of primary keys for advertisement
+ *              documents in the "ads" collection.
+ * Author: Kira Toal
+ * Date: July 20, 2020
+ */ 
+public class StateCollectionBuilder {
+
+  private static final Set<String> VALID_GEO_TARGETS = new HashSet<>(Arrays.asList("Alabama", "Alaska", 
+          "Arizona", "Arkansas", "California", "Colorado", "Connecticut", "Delaware", "Florida", "Georgia", 
+          "Hawaii", "Idaho", "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky", "Louisiana", "Maine", 
+          "Maryland", "Massachusetts", "Michigan", "Minnesota", "Mississippi", "Missouri", "Montana", 
+          "Nebraska", "Nevada", "New Hampshire", "New Jersey", "New Mexico", "New York", "North Carolina", 
+          "North Dakota", "Ohio", "Oklahoma", "Oregon", "Pennsylvania", "Rhode Island", "South Carolina", 
+          "South Dakota", "Tennessee", "Texas", "Utah", "Vermont", "Virginia", "Washington", "West Virginia", 
+          "Wisconsin", "Wyoming", "United States"));
+  private static final String PATH_TO_SERVICE_ACCOUNT = "./serviceAccountKey.json"; 
+  private static final String DATABASE_URL = "https://step9-2020-capstone.firebaseio.com"; 
+  private static final String MAIN_COLLECTION = "ads";
+  private static Firestore db; 
+
+  public static void initializeApp() throws Exception {
+    // Set account and build options.
+    FileInputStream serviceAccount = new FileInputStream(PATH_TO_SERVICE_ACCOUNT);
+    FirebaseOptions options = new FirebaseOptions.Builder()
+        .setCredentials(GoogleCredentials.fromStream(serviceAccount))
+        .setDatabaseUrl(DATABASE_URL)
+        .build();
+    
+    // Initialize app.
+    if(FirebaseApp.getApps().isEmpty()) {
+      FirebaseApp.initializeApp(options);
+    }
+    db = FirestoreClient.getFirestore();
+  }
+
+  public static void updateStateCollection(String state, List<QueryDocumentSnapshot> documents) throws Exception {
+    // Initialize a HashMap to organize ad data into documents.
+    Map<String, ArrayList> advertiserToAdIds = new HashMap<>();
+
+    // Create a document for every advertiser.
+    for (QueryDocumentSnapshot document : documents) {
+      Ad ad = document.toObject(Ad.class); // Convert document to Ad to use methods like .getAdvertiser().
+      String key = ad.getAdvertiser();
+      ArrayList<String> value = new ArrayList<String>(); 
+      if (advertiserToAdIds.containsKey(key)) {
+        value = advertiserToAdIds.get(key);
+      }
+      value.add(ad.getId());
+      advertiserToAdIds.put(key, value);
+    }
+
+    // Add advertiser documents to the corresponding state collection.
+    for (String key : advertiserToAdIds.keySet()) {
+      Map<String, Object> data = new HashMap<>();
+      data.put("ads", advertiserToAdIds.get(key));
+      db.collection(state.toLowerCase()).document(key).set(data, SetOptions.merge());
```

Comment:
This should really be in a sub collection called `states` where each document in `states` is a `state` that has a collection called `advertisers` or something. Don't want to pollute the main DB collection with 50+ `collections` that are easily grouped.